Add a new gettimespan function for simple durations

lburg opened the pull request:
The library was missing a simple way to express durations, what do you think about this simple implementation?

dbaty commented on getconf/base.py:
```diff
@@ -275,6 +276,34 @@ def getfloat(self, key, default=0.0, doc=''):
+            return None
+
+        conversion = {'d': 'days', 'h': 'hours', 'm': 'minutes', 's': 'seconds'}
+        value, duration = value[:-1], value[-1:]
```
I would rename `duration` as `unit`. Ditto in the log message below.

dbaty commented on getconf/base.py:
```diff
@@ -275,6 +276,34 @@ def getfloat(self, key, default=0.0, doc=''):
+                "%s is not a valid duration for the key %s (available values are %s).",
+                repr(duration),
+                key,
+                ','.join(conversion.keys()),
```
`', '` so that we have a space between values?

Minor: `available` -> `acceptable`... but maybe it's just me! :)

RemiCardona commented on tests/test_base.py:
```diff
@@ -374,6 +375,58 @@ def test_getfloat_defaults_raises(self):
+        self.assertRaises(AssertionError, lambda: getter.gettimespan('test', False))
+        self.assertRaises(AssertionError, lambda: getter.gettimespan('test', 42))
+        self.assertRaises(AssertionError, lambda: getter.gettimespan('test', 4.2))
+        self.assertRaises(AssertionError, lambda: getter.gettimespan('test', (1, )))
```
No need for a lambda, `assertRaises()` allows *args and **kwargs
```python
self.assertRaises(AssertionError, getter.gettimespan, 'time', <wrong value>)
```

RemiCardona commented on tests/test_base.py:
```diff
@@ -374,6 +375,58 @@ def test_getfloat_defaults_raises(self):
+        getter = getconf.ConfigGetter('TESTNS')
+        self.assertEqual(datetime.timedelta(), getter.gettimespan('test'))
+        self.assertEqual(datetime.timedelta(days=100), getter.gettimespan('foo', '100d'))
+        self.assertEqual(None, getter.gettimespan('test', None))
```
Use `self.assertIsNone()`

RemiCardona commented on getconf/base.py:
```diff
@@ -275,6 +276,34 @@ def getfloat(self, key, default=0.0, doc=''):
+        except ValueError:
+            logger.exception("Unable to cast %s as float for the key %s.", repr(value), key)
+            raise
+        except KeyError:
```
`KeyError` should not leak out to the caller, it's an implementation detail here. Given the other methods, it should probably be converted to a `ValueError`.

bors-ltd commented on tests/test_base.py:
```diff
@@ -364,15 +365,67 @@ def test_getfloat_defaults(self):
+        self.assertRaises(AssertionError, getter.getfloat, 'test', False)
+        self.assertRaises(AssertionError, getter.getfloat, 'test', 42)
+        self.assertRaises(AssertionError, getter.getfloat, 'test', '4.2')
+        self.assertRaises(AssertionError, getter.getfloat, 'test', (1, ))
```
Merci d'avoir nettoyé la planète de tous ces lambda. :clap:

Même si c'est un peu hors-sujet dans cette PR. :-)

RemiCardona replied:
👍 on removing all those lambdas, but the part that's not relevant to the current PR should be done in a separate commit

bors-ltd commented on getconf/base.py:
```diff
@@ -275,6 +276,34 @@ def getfloat(self, key, default=0.0, doc=''):
             logger.exception("Unable to cast %s as float for the key %s.", repr(value), key)
             raise
 
+    def gettimespan(self, key, default='0d', doc=''):
```
Une raison pour ne pas l'avoir appelé `gettimedelta` puisque c'est son type de retour, tout comme `getstr` renvoit un `str`, `getint` un `int`, etc. ?

lburg commented:
Juste force pushed new commits. I renamed the function to `gettimedelta`, I thought `span` was prettier but consistency is more important. Split into two commits as suggested.
